GC safepoint and transition support

yuyichao opened the pull request:
This is a work in progress to make our GC a stop the world GC rather than a wait for the world (and maybe dead lock?) GC _(ok that probably doesn't sound too exciting...)_. ~~The first two commits are #14181 which I made as a general improvement to our finalizers handling while trying to figure out how to synchronize finalizers and the GC with threading for this PR.~~ (merged)

This is still missing the very important GC transition and safepoint support in codegen but I'm opening this as RFC now to make sure that the GC synchronization schematics I'm following makes sense.

I described the schmatics in [the comment in `gc.c`](https://github.com/JuliaLang/julia/commit/a08e6ba38321c46349b34da207c183096aee0240#diff-31b93f4272eaeede3738531e90317e0cR41). To summarize, we need GC safe point and GC transitions so that the GC doesn't have to wait for other threads forever (especially if the thread doing the GC is holding a lock). The safepoint is implemented using a volatile load from a special page which we `mprotect` to `PROT_NONE` during a GC and catch the SegFault in the signal handler. (This is similar to how the Java HotSpot GC does it.) The GC also makes sure (with a combination of lock and safe point) that no other thread can run the GC at the same time or modifying GC related data structures when the GC is running.
#### TODO/request for help:
- The segfault handler for OSX is missing. It seems that I can't just wait in the signal handler since it's running on a different thread processing the messages from the kernel in series and it is necessary to do some low level stuff to make sure the thread can return to it's previous state after waiting for the GC to finish. Doing this on a platform that I can't test locally is beyond what I can confidently do so I need help from someone who's more familiar with this to implement it.
- Codegen support:
  
  I'm currently planning to emit a GC transition around each critical region (store to object or stack) and allow the GC to run everywhere else. We can do a optimization pass to merge unnecessary GC transitions after the whole function is generated. I'm still deciding what I can assume about GC awareness for a function call and what's the convension of changing the GC state in a function. Note that this shouldn't require the LLVM statepoint support so it should work on LLVM 3.3 as well.

c.c. @vtjnash @kpamnany @carnaval @Keno 

Close https://github.com/JuliaLang/julia/issues/13380
Close https://github.com/JuliaLang/julia/issues/14343
Close https://github.com/JuliaLang/julia/issues/10317
##### TODO for this PR:
- [x] SegFault handler for OSX
- [x] Rename `managed` / `unmanaged` to `unsafe` / `safe`
- [x] Run finalizers after the GC is finished and allowing GC to run in finalizers ~~(_Still need to update the comment in `gc.c`_)~~ done.
- [x] ~~Easier way to breakpoint on actual segfault~~ (`jl_critical_error` should be good enough)
##### TODO that can be done in future PRs:
- [ ] Make sure the runtime satisfies the GC state transition requirements
- [ ] Insert safepoints and state transitions in codegen
- [ ] Optimization pass on state transitions
- [ ] Fix GC debugging code (stack grub) (~~requires https://github.com/JuliaLang/julia/pull/14473~~ merged)

yuyichao committed:
GC safepoint and GC transition

vtjnash commented on src/gc.c:
```diff
@@ -2277,10 +2435,51 @@ JL_DLLEXPORT void jl_gc_collect(int full)
 }
 
+JL_DLLEXPORT void jl_gc_collect(int full)
+{
```
should there be a statepoint here?

vtjnash commented on src/gc.c:
```diff
@@ -2277,10 +2435,51 @@ JL_DLLEXPORT void jl_gc_collect(int full)
+JL_DLLEXPORT void jl_gc_collect(int full)
+{
+    if (!is_gc_enabled || jl_in_gc)
+        return;
```
should there be a statepoint here?

yuyichao replied:
No. Multiple threads entering the GC is handled separately below to avoid running the GC multiple times.

vtjnash replied:
i didn't realize jl_in_gc was thread-local now. it seems like that flag should never be true here now in that case?

to return to the original purpose of this flag, it should perhaps be renamed `jl_in_finalizers` and moved to only protecting the finalizer queue from task switches? the original intent of this flag was to disallow finalizers from descheduling the thread and switching to running other tasks at unexpected yield points (any allocation).

yuyichao replied:
It could be true in finalizer.

> it should perhaps be renamed `jl_in_finalizers`

Yeah, that's effectively what it is for now.

JeffBezanson commented on src/threading.c:
```diff
@@ -215,6 +215,12 @@ void ti_threadfun(void *arg)
+    // critical region. In general, the following part of this function
+    // shouldn't call any managed code without calling `jl_gc_managed_enter`
+    // first.
+    jl_gc_state_set_and_save(3, 0);
```
What does `3` mean?

yuyichao replied:
I originally meant unmanaged code but I don't think that distinction is important.

yuyichao replied:
(instead of pausing waiting for or running GC, which is `gc_state == 1`)

JeffBezanson commented:
This is great. I think we might be able to merge this fairly soon, since even without codegen support it appears to be strictly better than what we have now, where there is effectively a single GC statepoint (plus it doesn't work).

vtjnash commented on src/julia.h:
```diff
@@ -1446,9 +1465,46 @@ JL_DLLEXPORT JL_CONST_FUNC jl_tls_states_t *(jl_get_ptls_states)(void);
+// Add two arguments, one for the default value of `old_state`, one to satisfy
+// ISO C++11 standard.
+#define jl_gc_state_set_and_save(...)           \
+    jl_gc_state_set_and_save__(__VA_ARGS__, 0, 0)
```
`"Variadic macros are a new feature in C99. GNU CPP has supported them for a long time, but only with a named variable argument (‘args...’, not ‘...’ and __VA_ARGS__). If you are concerned with portability to previous versions of GCC, you should use only named variable arguments. On the other hand, if you are concerned with portability to other conforming implementations of C99, you should use only __VA_ARGS__."`

yuyichao replied:
I think the only question is what does MSVC support and IIRC it supports the `__VA_ARGS__` version in c++ mode. @tkelman

yuyichao replied:
And in the end this is just some trick to avoid having two names so can easily be removed if we can't support on all compilers.

tkelman replied:
```
d:/code/msys64/home/Tony/julia/src/gc.c(2480) : error C2660: 'jl_gc_state_set_an
d_save_' : function does not take 3 arguments
```

yuyichao replied:
... fine... I can just rename it.........

yuyichao commented:
Yes, it should be mostly safe to merge in it's current state. The missing part for the state point is OSX support and I'm hoping **someone** more familiar with x64 calling convension and OSX signal handling can implement that ;-).

I'm currently working on settling on a GC state transition convention and make sure the the runtime follows that convension (which is time consuming since I need to go through all the functions in `src/`).

As I mentioned in other threads, there's a few other GC fixes that are kind of independent with this PR including locking for write barrier and scanning thread stack correctly. Those should be relatively easy to fix, I'll probably submit a PR for those this weekend but anyone else should feel free to fix those.

JeffBezanson commented:
IIUC, the purpose of the GC state transitions is to allow GC inside unmanaged code? I don't really like the idea of needing to insert lots of explicit state transitions all over. I would rather interrupt threads asynchronously and do conservative register scanning.

vtjnash commented on src/threading.c:
```diff
@@ -237,11 +243,18 @@ void ti_threadfun(void *arg)
+                // TODO: before we support getting return value from
+                //       the work, we don't need to enter GC managed region
+                //       when starting the work.
+                int8_t state = jl_gc_managed_enter();
```
should this `state` be added to the exception state (jl_enter / jl_leave / jl_rethrow)?

yuyichao replied:
This, it is on my uncommitted local changes.

vtjnash commented:
> The segfault handler for OSX is missing. It seems that I can't just wait in the signal handler since it's running on a different thread processing the messages from the kernel in series and it is necessary to do some low level stuff to make sure the thread can return to it's previous state after waiting for the GC to finish. Doing this on a platform that I can't test locally is beyond what I can confidently do so I need help from someone who's more familiar with this to implement it.

i believe the right approach is to return `MIG_NO_REPLY` from `catch_exception_raise` (which will leave the thread unscheduled), then call `thread_resume` to continue running it later after gc is finished.

yuyichao commented:
I agree that could be simpler if we have a conservative scan already. (The current approach isn't too bad though, and it's nice to see that some functions can actually be run simultaneously with the GC).

More importantly, I think it is still possible to pause a thread at a point that can cause corruption. (in the GC allocation functions, for example). I'm also not totally convinced that the GC can have a consistent view of the object if we pause the thread at any time. (e.g. what if we pause a thread in the middle of a buffer `realloc`). I'm probably a little pessimistic about this but I think it might be less code to add indeed but takes equal amount of effort to make sure we've covered all the cases (given how many GC roots we've missed in the past...)

yuyichao commented:
Another thing I'm worrying is that what would happen if the thread haven't got a change to rerun the segfault'd load before the next GC is triggered?

vtjnash commented:
it should re-trigger the statepoint, which would be fine. there might be a race condition with testing and setting `jl_gc_running` / `ptls->state` though

JeffBezanson commented:
Will this, for example, require switching gc states around every ccall? I would *not* be ok with that.

I think we could stage this by first just merging the new gc waiting mechanism, with minimal state points: only in jl_gc_collect and when entering or exiting threaded regions. Then the statepoints can be slow, we can leave the trickier parts for later, and in the meantime threading will be much more stable.

yuyichao commented:
> Will this, for example, require switching gc states around every ccall?

No

> I would not be ok with that.

Me neither.

The plan is to run in unmanaged mode by default and only add transitions around critical regions (write barrier and gc frame store) and then merge the critical regions if there's no code in between that has to be running in unmanaged mode. This way we only do the expensive transition when there's more expensive operations (note that the source of the store are usually also expensive).

The worst case is of course if you have alternating code that need to be managed and unmanaged and if the source is actually cheap (in which case the conservative scan should perform well). I did a small benchmark to check what the overhead would be with the code below (note that in this case we can hoist the transition out of the loop and only leave the safepoint check in the loop but it should be the worst case we could ever have).

The timing on my laptop shows that `f1` is ~10% faster then `f2` (no difference if I hoist the transition) when I'm only setting the field `a`. There's no measurable difference if I set both field.

```
type Obj{T}
    a::T
    b::T
end

function f1(n, p1, p2, obj, a, b)
    for i in 1:n
        obj.a = a
        # obj.b = b
    end
end

function f2(n, p1, p2, obj, a, b)
    for i in 1:n
        Base.llvmcall("""
                      store volatile i8 3, i8* %0
                      %3 = load volatile i64, i64* %1
                      ret void
                      """, Void, Tuple{Ptr{Int8},Ptr{Int64}},
                      Ptr{Int8}(p1), Ptr{Int64}(p2))
        obj.a = a
        # obj.b = b
        Base.llvmcall("""
                      store volatile i8 0, i8* %0
                      ret void
                      """, Void, Tuple{Ptr{Int8}}, Ptr{Int8}(p1))
    end
end

using Benchmarks

a = Ref(1)
b = Ref(1)
obj = Obj(a, b)
b1 = Ref(1)
b2 = Ref(1)
p1 = pointer_from_objref(b1)
p2 = pointer_from_objref(b2)

@show @benchmark f1(10, p1, p2, obj, a, b)
@show @benchmark f2(10, p1, p2, obj, a, b)
```

JeffBezanson commented:
Thanks, I understand it better now.

yuyichao committed:
Preserve GC state across task switch

yuyichao committed:
Make sure debuginfo won't dead lock during GC

yuyichao committed:
Avoid dead lock due to ti_threadgroup_join

yuyichao committed:
Temporary fix to make sure GC while holding a lock will not dead lock.

Add tests.

yuyichao committed:
Thread safe jl_gc_enable

yuyichao commented:
[Travis/osx](https://travis-ci.org/JuliaLang/julia/jobs/95148780) has the repl SIGINT issue. [Travis/lin64](https://s3.amazonaws.com/archive.travis-ci.org/jobs/95148779/log.txt) generates too much log compiling openblas (why is it doing that?) and got killed....

tkelman commented:
We can't use the PPA on Travis any more. Either we can use sudo-enabled workers that will fail the socket test or we can use no-sudo docker workers where we need to build and cache deps from source. There was a caching issue, should be worked out now I think. Once in a while we may need to clean out the cache under the Travis settings and have a build on master take longer to re-populate it.

tkelman commented:
It was actually an openblas compilation failure, just got logged in a slightly out-of-order way.
```
collect2: ld terminated with signal 9 [Killed]
make[2]: *** [xscblat3] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [tests] Error 2
make: *** [build/openblas-53e849f4fcae4363a64576de00e982722c7304f9/libopenblas64_.so] Error 1
make: *** Waiting for unfinished jobs....
```
Travis terminates your job if you have no output for 10 minutes. So I'm building deps outputting just stdout to a log file, but leaving stderr printing as normal since we have enough warnings to keep the job alive but not too many warnings to go over the 4 MB max log length at which Travis also terminates your job. So the stderr comes first, then if anything goes wrong in building the deps I print the stdout from the log file. Suggestions welcome, maybe we could do just the tail of the stdout log file.

yuyichao commented:
That makes more sense now. I was wondering why does other builds and in particular the build that poopulate the cache doesn't have this problem....

JeffBezanson commented on src/jlapi.c:
```diff
@@ -318,6 +318,26 @@ JL_DLLEXPORT jl_value_t *(jl_typeof)(jl_value_t *v)
+    jl_gc_managed_leave(state);
+}
+
+JL_DLLEXPORT int8_t (jl_gc_unmanaged_enter)(void)
```
I would rather call this `gc_safe` --- you call it that in your comments elsewhere, and I think it's clearer.

yuyichao replied:
Sure. I'm fine with any names.

yuyichao replied:
Rename done. Also added some comment for different values of the `gc_state`.

yuyichao commented:
I was testing various (evil) multithread workload on this branch and it seems that all of the SegFault I've got are due to interaction with the GC rather than the GC synchronization itself (Ref https://github.com/JuliaLang/julia/pull/14361, https://github.com/JuliaLang/julia/pull/14366, https://github.com/JuliaLang/julia/pull/14345, https://github.com/JuliaLang/julia/pull/14307, https://github.com/JuliaLang/julia/pull/14301). Therefore, I think this PR is ready to merge apart from the OSX segfault handler. As decided above, this PR only includes the synchronization mechanism and minimum amount of safepoints. I'm working on the runtime and codegen support on different branches and will open another PR when it's ready.

Breif notes for debugging threading version in the future. Due to the use of segfault as GC safepoint (which IMHO is the lowest mutator overhead mechanism that works with code loaded from disk) there can be many segfault in normal operation if multiple mutator threads are running (shouldn't be an issue if only one of them is running). Therefore, it is recommanded to ignore segfault in the debugger. The GDB command for this is `handle SIGSEGV noprint nostop pass` and I believe the lldb equivalent is `pro hand -p true -s false -n false SIGSEGV`. If you need to catch an actual segfault, set break point on `jl_critical_error` (or `sigdie_handler` also works on linux) should do it. I'm not sure how will this affect OSX yet (not sure at which stage the debugger captures the segfault).

@Keno and @vtjnash have agreed to implement the OSX version of the safepoint. Here's a breif description of the requirements.

1. The thread will recieve a segfault if it hits a safepoint when the GC is running. The signal handler
    needs to put the thread to wait (ideally suspend it) until the GC is finished. Whether the GC is
    running is indicated by the `jl_gc_running` variable (feel free to un`static` it). It would be nice
    if we can make the thread also run GC code in the future but we currently only need on thread for
    the GC.
2. When the GC is done, it will set `jl_gc_running` to `0`. We may need to send another message
    to the signal thread so that it can wake the threads it suspended up. (Alternatively we could also
    lock the suspended thread list and let the GC thread wake up the threads).
3. Of course, implement this in a none-racy way =).

@tkelman (or other windows users) Could you please check if this is working properly on windows? Passing the `threads` test (which includes a test that dead locks on master now) should be a good start. If you are ambitious enough, you can also try [this script](https://github.com/yuyichao/explore/blob/d62ec5717e6f2362bdefb26f1e184bcd0227ee29/julia/gc_thread/gc-thread.jl). THX.

yuyichao commented:
Actually I think I'll add the notes about debugging to the dev manual once I know what's the situation with OSX.

tkelman commented:
Will have to check whether this regressed on master or within this PR, but win64 with threading enabled:
```
$ gdb -q --args /home/Tony/julia/usr/bin/julia-debug.exe -C x86-64 --output-ji `cygpath -w /home/Tony/julia/usr/lib/julia/inference0.ji` -f coreimg.jl
Reading symbols from /home/Tony/julia/usr/bin/julia-debug.exe...done.
(gdb) r
Starting program: /home/Tony/julia/usr/bin/julia-debug.exe -C x86-64 --output-ji D:\\cygwin64\\home\\Tony\\julia\\usr\\lib\\julia\\inference0.ji -f coreimg.jl
[New Thread 10148.0x201c]

Program received signal SIGSEGV, Segmentation fault.
0x000000006ea638ff in jl_gc_safepoint () at /home/Tony/julia/src/julia.h:72
72          *jl_gc_signal_page;
(gdb) bt
#0  0x000000006ea638ff in jl_gc_safepoint () at /home/Tony/julia/src/julia.h:72
#1  0x000000006ea6950a in jl_gc_enable (on=0) at /home/Tony/julia/src/gc.c:2141
#2  0x000000006ea340c5 in _julia_init (rel=JL_IMAGE_JULIA_HOME) at /home/Tony/julia/src/init.c:544
#3  0x000000006ea35c48 in julia_init (rel=JL_IMAGE_JULIA_HOME) at /home/Tony/julia/src/task.c:273
#4  0x0000000000402f80 in wmain (argc=1, argv=0x3a8cb0, envp=0x3ab120)
    at /home/Tony/julia/ui/repl.c:625
#5  0x000000000040140c in __tmainCRTStartup ()
    at /usr/src/debug/mingw64-x86_64-runtime-4.0.2-1/crt/crtexe.c:329
#6  0x000000000040153b in mainCRTStartup ()
    at /usr/src/debug/mingw64-x86_64-runtime-4.0.2-1/crt/crtexe.c:212
```

yuyichao commented:
@tkelman This is expected and is how the GC safepoint works. See my debugging notes above. Do you see segfault without attaching a debugger?

tkelman commented:
Ah, my bad. Yes it segfaults without a debugger.

yuyichao commented:
Hmm. yeah, the backtrace is strange. It should trigger a safepoint this early in initialization. I'm wondering if there's any problem with allocating the page. Could you please try to replace the `MEM_RESERVE` in `jl_gc_signal_init` (not `malloc_page`) with `MEM_COMMIT`? That is what [this blog](http://ntcoder.com/bab/2009/02/14/how-to-create-protected-readonly-pages/) uses. If that doesn't work, maybe try to change the `PAGE_READONLY` in this function and `jl_gc_signal_end` with `PAGE_READWRITE`?

tkelman commented:
Changing `MEM_RESERVE` to `MEM_COMMIT` seems to work.

tkelman commented on src/gc.c:
```diff
@@ -2317,7 +2430,53 @@ JL_DLLEXPORT void jl_gc_collect(int full)
+#else
+        // For single thread, GC should not call itself (in finalizers) before
+        // setting jl_gc_running to false so this should never happen.
+        assert(0 && "GC synchronization failure");
```
this fails in an msvc build with threading disabled, so I guess the atomics aren't right there?

yuyichao replied:
What exactly is the the failure? Is that a compile time error or assertion trigger at runtime?

Could you please paste the compile error or the backtrace of the assertion?

tkelman replied:
It's during bootstrap
```
make[1]: Entering directory `/d/code/msys64/home/Tony/julia'
 cd /d/code/msys64/home/Tony/julia/base && /d/code/msys64/home/Tony/julia/usr/bi
n/julia.exe -C x86-64 --output-ji /d/code/msys64/home/Tony/julia/usr/lib/julia/i
nference0.ji -f coreimg.jl
essentials.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
range.jl
expr.jl
error.jl
bool.jl
number.jl
int.jl
operators.jl
pointer.jl
abstractarray.jl
array.jl
hashing.jl
nofloat_hashing.jl
functors.jl
reduce.jl
intset.jl
dict.jl
iterator.jl
inference.jl
Assertion failed: 0 && "GC synchronization failure", file d:/code/msys64/home/To
ny/julia/src/gc.c, line 2459
make[1]: *** [/d/code/msys64/home/Tony/julia/usr/lib/julia/inference0.ji] Error
3
make[1]: Leaving directory `/d/code/msys64/home/Tony/julia'
make: *** [julia-inference] Error 2
```
not sure about getting a backtrace in an msvc build, requires windbg

yuyichao replied:
> not sure about getting a backtrace in an msvc build, requires windbg

Maybe use our own unwinder then? Something like calling `gdbbacktrace()` before the assertion. (Or when I was lazy and don't want to figure out what to use, I'll probably do a `*(volatile int*)NULL;` ;-p )

yuyichao replied:
@tkelman You are right. The definition of the atomic macros for windows are wrong. Will try to push a fix.

tkelman commented on src/signals-win.c:
```diff
@@ -169,6 +169,13 @@ static LONG WINAPI _exception_handler(struct _EXCEPTION_POINTERS *ExceptionInfo,
             case EXCEPTION_ACCESS_VIOLATION:
+#ifdef JULIA_ENABLE_THREADING
+                if (ExceptionInfo->ExceptionRecord->ExceptionInformation[1] ==
+                    jl_gc_signal_page) {
```
msvc failure
```
d:\code\msys64\home\tony\julia\src\signals-win.c(174) : error C2446: '==' : no c
onversion from 'volatile size_t *' to 'ULONG_PTR'
        There is no context in which this conversion is possible
d:\code\msys64\home\tony\julia\src\signals-win.c(174) : error C2040: '==' : 'ULO
NG_PTR' differs in levels of indirection from 'volatile size_t *'
```

yuyichao committed:
Match the behavior of the atomic intrinsics on different compilers.

yuyichao commented:
Hopefully all the issues mentioned above are fixed. I made the schematics match the MSVC one since GCC also provide those but not the other way around...

This could also explain why MSVC threading wasn't working, the lock basically won't work, and it might worth trying that again.

tkelman commented:
Nice, looks like this does fix the bootstrap freeze when threading is enabled with MSVC.

Should some of the devdocs and debugging tips be adjusted for the new gdb segfault instructions?

yuyichao commented:
> Should some of the devdocs and debugging tips be adjusted for the new gdb segfault instructions?

[Yes](https://github.com/JuliaLang/julia/pull/14190#issuecomment-163979413). Jameson confirmed that this shouldn't be an issue on OSX since our handler get the message before ptrace so I'll update the doc soon.

yuyichao committed:
Update debugging tips for gc safepoint and other minor changes.

yuyichao commented:
Debugging tips updated. Hopefully I've got the gramma and rst right............

yuyichao committed:
Lock based safepoint for mach.

Also fix other TLS references on the signal handling thread for OSX. Add
comment to the functions it uses so that we don't easily break it.

yuyichao commented:
I added the OSX signal handler (https://github.com/JuliaLang/julia/commit/98b995e86377b5d0155973b7588176630741269f) using a lock for synchronization, which I think is probably the best solution anyway since it doesn't need to go through the kernel when the GC is finished like it would if we do that by sending a message to the signal handler thread.

However, when enabling it on travis, I'm getting a SIGBUS in the threads test and I'm not sure what causes that.

@Keno @vtjnash

yuyichao commented:
The OSX segfault handler should be fixed now. The issue was that we didn't register the segfault handler for all the threads we started on OSX..........

See https://ci.appveyor.com/project/yuyichao/julia and https://travis-ci.org/yuyichao/julia/builds/98020723 for the CI with threading enabled (on a branch including other fixes)

I've checked off all the todo items for this PR and removed the `[WIP, RFC]` tag. This PR should be ready for final review now. =)

@JeffBezanson @vtjnash

JeffBezanson commented on src/julia.h:
```diff
@@ -1460,16 +1491,50 @@ JL_DLLEXPORT JL_CONST_FUNC jl_tls_states_t *(jl_get_ptls_states)(void);
+}
+#define jl_gc_unsafe_enter() jl_gc_state_save_and_set(0)
+#define jl_gc_unsafe_leave(state) jl_gc_state_set((state), 0)
+#define JL_RETURN_NULL_IF_UNMANAGED() do {       \
```
What is this for?

yuyichao replied:
When a thread is in a gc safe state, any newly allocated return values are invalid. This is used in runtime functions that allocates new return values so that a invalid return value can be immediately detected.

yuyichao replied:
I realized that this is probably only needed on the branch that I'm adding the runtime support and I also forgot to rename it. Should I move this to a later commit or should I just rename it to SAFE instead.

yuyichao replied:
Renamed

JeffBezanson commented on src/jlapi.c:
```diff
@@ -318,6 +318,26 @@ JL_DLLEXPORT jl_value_t *(jl_typeof)(jl_value_t *v)
+    jl_gc_unsafe_leave(state);
+}
+
+JL_DLLEXPORT int8_t (jl_gc_safe_enter)(void)
```
What's the difference between safe_enter and unsafe_leave? Seems to imply you can be in a state that's neither safe nor unsafe?

yuyichao replied:
Well, you are either safe and unsafe but you might not know which one your are in when entering the function. This is important so that when we add codegen support, we can move a GC transition across a runtime function call and merge it with another transition.

The `safe` / `unsafe` `enter` function set the state to safe or unsafe and return the current state, the `safe` / `unsafe` `leave` function then set it back.

yuyichao commented:
Any other comments?

yuyichao commented:
Rebased on top of #14501 . Also fixes other TLS references in the signal handling thread on OSX.

JeffBezanson commented on src/debuginfo.cpp:
```diff
@@ -194,7 +194,9 @@ class JuliaJITEventListener: public JITEventListener
     virtual void NotifyFunctionEmitted(const Function &F, void *Code,
                                        size_t Size, const EmittedFunctionDetails &Details)
     {
+        int8_t gc_state = jl_gc_state_save_and_set(3);
```
This `3` should have a name.

yuyichao replied:
Done. Also for the `1`.

I also change it to `2` instead since I don't think we need it to be a bit flag (which was why I use `3` instead of `2` to begin with).

JeffBezanson commented on src/julia.h:
```diff
@@ -1462,16 +1493,52 @@ JL_DLLEXPORT JL_CONST_FUNC jl_tls_states_t *(jl_get_ptls_states)(void);
+#define jl_gc_unsafe_leave(state) jl_gc_state_set((state), 0)
+// Return NULL when we are in GC safe mode and the return value can be free'd
+// by GC on another thread
+#define JL_RETURN_NULL_IF_GCSAFE() do {          \
```
Does it make sense to remove this macro from this PR since it's not used? Also I'm not sure how I feel about `return` inside a macro...

yuyichao replied:
Also done. I also don't really like `return` inside a macro although I sometimes treat error checking macros with a `return` in the name as the only exception.

I used a macro here since I want to make sure the non-threading version has neither the volatile load and the branch. I just realized that I could just define `jl_gc_state` as `0` on non-threading and branch on that for `NULL` return.

yuyichao committed:
Move execution of finalizes out of `jl_gc_running = 1`.
Rename `jl_in_gc` to `jl_in_finalizer`
Allow running GC in finalizers

yuyichao commented:
Rebased and CI all pass. Given that there's only very minor comments and updates in the past two weeks and people have probably back to work now, I'll merge this sometime later this week. It is very hard to make further progress on threading without merging the current PR's.

JeffBezanson commented:
Yes, let's merge this. Amazing work @yuyichao.

I have a couple more questions that are not urgent and don't block merging:
- Where is the initial gc state for the master thread set? I expected to see something in `ti_initthread`.
- It looks like waiting for gc uses busy waiting? Can that be fixed at some point?

yuyichao commented:
> Where is the initial gc state for the master thread set? I expected to see something in ti_initthread.

I assumed that the tls buffer is 0 initialized since it's a static variable in C. Should we explicitly initialize that?

> It looks like waiting for gc uses busy waiting? Can that be fixed at some point?

Right, that follows from what we use for our runtime lock. I don't like this busy loop either and I think both should be fixed (especially for the cases we are doing expensive stuffs in the runtime). FWIW, the pthread lock performs better than our spin lock on my laptop when I was benchmarking https://github.com/JuliaLang/julia/pull/14451.

JeffBezanson commented:
> Should we explicitly initialize that?

Yes I think it's clearer, just for the sake of reading the init code and wanting to know the initial state.

yuyichao commented:
Done https://github.com/JuliaLang/julia/pull/14565
